### ISSUE TITLE
Issue with two-valued GL version numbers

### ISSUE THREAD

**Issue by oscargus** [pyglet/pyglet#36]
(Reopening https://bitbucket.org/pyglet/pyglet/issues/228/issue-with-two-valued-gl-version-numbers as I didn't have a bitbucket account and therefore missed the request for more information)

A traceback can be found in https://github.com/sympy/sympy/issues/16537#issuecomment-487540560 

I am pretty sure that it is exactly the same error that I got. (I also realize that it would have been easier to at least point you to where the error happened in the original report. Sorry about that.)

**Comment by benmoran56** [comment]
Hi @oscargus, 

Thanks for getting back to this. Now that you're here, we can test a fix and make sure it works on both two and three value version numbers. I'll see if I can put a test together soon.

**Comment by benmoran56** [comment]
This should now be fixed in master. I ended up changing the way versions are queried, so we're no longer relying on string parsing. If you could give it a try that would be great. If you've installed via pip, you can upgrade to master with something like: 

`pip install --upgrade git+https://github.com/pyglet/pyglet.git --user`

**Comment by benmoran56** [comment]
It turns out that this change breaks OSX, which does not seem to support GL_MAJOR_VERSION / GL_MINOR_VERSION querying. :disappointed: 

Looks like we'll have to revert to string parsing for now, but we can fix that instead. Please go ahead and give it a test anyway if you're not on OSX.

**Comment by oscargus** [comment]
Thanks for the rapid reply and action! I'll give it a go on my office machine (where it happens, Linux) when I'm back at the office. (I realize that the IT department has changed OS and probably drivers since then, but let us see what happens...)

Maybe using `LooseVersion` (or `StrictVersion`, I do not really know the difference) from `distutils.version` can be an option. Then one can do thing like:

`if LooseVersion(git_ver) < LooseVersion(minimal):`

with string representation of the versions.

I noted that you have temporarily removed some tests in master. Will try to get to the office tomorrow then. Need to go in a few days this week anyway. :-)

**Comment by benmoran56** [comment]
Is there any chance of getting an example of the version string? 
On my machine, for example, it displays as:  **"3.0 Mesa 19.1.3"** for example. 
You can query this via: 
```$ glxinfo |grep "OpenGL version"```
```OpenGL version string: 3.0 Mesa 19.1.3```

That would be very useful. I'll post on the sympy issue as well with the same question.

**Comment by oscargus** [comment]
OpenGL version string: 4.6.0 NVIDIA 430.26

The error remains (without updating). Will try with the new version now.

**Comment by oscargus** [comment]
Now I get

```
Exception in thread Thread-20:
Traceback (most recent call last):
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/nobackup/data/sympy/sympy/plotting/pygletplot/managed_window.py", line 56, in __event_loop__
    clock.set_fps_limit(self.fps_limit)
AttributeError: 'Clock' object has no attribute 'set_fps_limit'
```

which comes from SymPy. Will have to look into that first.

**Comment by oscargus** [comment]
It was easy to fix. No error message on close. Thanks!

**Comment by benmoran56** [comment]
Thank you for the version string. Just to confirm:  
1. It crashes with pyglet 1.4.1 installed from pip.
2. It does not crash with the latest code from master.

As I mentioned, we need to go back to parsing the vendor string for the version, because OSX doesn't support the new method in master. Now that I have an example string that breaks it, this should be easy to fix.

**Comment by benmoran56** [comment]
As a quick update, I have reverted, and then put in a new fix for the master branch. If possible, could you please test it again to see if it still works?

**Comment by oscargus** [comment]
1. Not sure. I think I had an older version than 1.4.1 installed.
2. Correct. The master from this morning worked without any error.

Will try with 1.4.1 and new master tomorrow.

**Comment by einarf** [comment]
This can be improved in 2.0:

https://www.khronos.org/opengl/wiki/OpenGL_Context#Context_information_queries

## OpenGL version number

To get the OpenGL major and minor version numbers, you can call these functions:

```
glGetIntegerv(GL_MAJOR_VERSION, *);
glGetIntegerv(GL_MINOR_VERSION, *);
```

These are **available on OpenGL version 3.0 and above contexts**. If those are not available, you can use this instead

**Comment by benmoran56** [comment]
Already done :+1: 

I had actually already merged that commit into the shaders branch, so I'll just clean it up a bit more.

**Comment by oscargus** [comment]
Now I've tried.

I still get an error with the pip 1.4.1.

For the new master, I also get an error:
```
Traceback (most recent call last):
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/nobackup/data/sympy/sympy/plotting/pygletplot/managed_window.py", line 72, in __event_loop__
    super(ManagedWindow, self).close()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/window/xlib/__init__.py", line 468, in close
    self.context.destroy()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/xlib.py", line 371, in destroy
    super(XlibContext13, self).destroy()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/base.py", line 374, in destroy
    gl._shadow_window.switch_to()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/window/xlib/__init__.py", line 488, in switch_to
    self.context.set_current()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/xlib.py", line 354, in set_current
    super(XlibContext13, self).set_current()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/base.py", line 333, in set_current
    gl_info.set_active_context()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 99, in set_active_context
    if self.have_version(3):
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 171, in have_version
    imajor, iminor, irelease = [int(v) for v in ver.split('.', 3)[:3]]
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 171, in <listcomp>
    imajor, iminor, irelease = [int(v) for v in ver.split('.', 3)[:3]]
ValueError: invalid literal for int() with base 10: ''
```

**Comment by oscargus** [comment]
Not sure if this helps, but for what it is worth:

```
In [8]: gl.GL_MAJOR_VERSION
Out[8]: 33307

In [9]: gl.GL_MINOR_VERSION
Out[9]: 33308

In [10]: gl.GL_VERSION
Out[10]: 7938

In [11]: gl.GL_VERSION_4_2
Out[11]: 1

In [12]: gl.GL_VERSION_4_1
Out[12]: 1

In [13]: gl.GL_VERSION_4_0
Out[13]: 1
```

**Comment by benmoran56** [comment]
Rats. The "base 10" error seems to be stemming from trying to int cast a space. If possible, can you modify your local "pyglet.gl.gl_info" module, and replace the "have_version" method with the following? 
```
    def have_version(self, major, minor=0, release=0):
        """Determine if a version of OpenGL is supported.

        :Parameters:
            `major` : int
                The major revision number (typically 1 or 2).
            `minor` : int
                The minor revision number.
            `release` : int
                The release number.
                :deprecated: No longer used

        :rtype: bool
        :return: True if the requested or a later version is supported.
        """

        if not self.have_context:
            warnings.warn('No GL context created yet.')
        if 'None' in self.version:
            return False
        ver = '%s.0.0' % self.version.strip().split(' ', 1)[0]
        print("ver: [{}], version: [{}]".format(ver, self.version))
        imajor, iminor, irelease = [int(v) for v in ver.split('.', 3)[:3]]
        return (imajor > major or
                (imajor == major and iminor >= minor) or
                (imajor == major and iminor == minor))
```

The print statement in there *should* show us exactly what the string looks like internally, before we try to do the "int(v) ...." list comprehension on it. 

Thanks again for your assistance with troubleshooting!

**Comment by oscargus** [comment]
Hmm, it seems like there may be an issue with SymPy... (I also realized that now I have a three-digit version number, but there used to be an issue with it earlier, I believe, if nothing else...)

```
In [4]: Plot(x*y**3-y*x**3)
ver: [4.6.0.0.0], version: [4.6.0 NVIDIA 430.26]
ver: [4.6.0.0.0], version: [4.6.0 NVIDIA 430.26]
Out[4]: [0]: -x**3*y + x*y**3, 'mode=cartesian'ver: [4.6.0.0.0], version: [4.6.0 NVIDIA 430.26]
ver: [.0.0], version: []
Exception in thread Thread-22:
Traceback (most recent call last):
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/nobackup/data/anaconda3/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/nobackup/data/sympy/sympy/plotting/pygletplot/managed_window.py", line 71, in __event_loop__
    super(ManagedWindow, self).close()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/window/xlib/__init__.py", line 468, in close
    self.context.destroy()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/xlib.py", line 371, in destroy
    super(XlibContext13, self).destroy()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/base.py", line 374, in destroy
    gl._shadow_window.switch_to()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/window/xlib/__init__.py", line 488, in switch_to
    self.context.set_current()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/xlib.py", line 354, in set_current
    super(XlibContext13, self).set_current()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/base.py", line 333, in set_current
    gl_info.set_active_context()
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 99, in set_active_context
    if self.have_version(3):
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 172, in have_version
    imajor, iminor, irelease = [int(v) for v in ver.split('.')[:3]]
  File "/home/oscgu95/.local/lib/python3.7/site-packages/pyglet/gl/gl_info.py", line 172, in <listcomp>
    imajor, iminor, irelease = [int(v) for v in ver.split('.')[:3]]
ValueError: invalid literal for int() with base 10: ''
```

**Comment by oscargus** [comment]
Or rather: I had a two-digit version number earlier and got that error. So I believed that it was the cause of the error based on the message.

Sorry about the hassle. I believe that this can be closed.

**Comment by oscargus** [comment]
If
https://github.com/pyglet/pyglet/blob/934c5d033f20240ba48e07fe5ab3c01a657ab722/pyglet/gl/gl_info.py#L99
is changed to
```
            if self.version and self.have_version(3):
```
it gives no error.

Maybe the reason is that the SymPy rendering is written in a bad way, not sure. It seems like it is basically calling close on the window so I cannot really understand why the GL information is gone. 

It may be a change only helpful for SymPy, but I cannot really see any negative side effects of it.

**Comment by benmoran56** [comment]
Hi @oscargus, 

Not a problem! 
Looks like a threading issue. At least now we know what was going on - the version issue was just blank, potentially due to the destruction of the window and/or context in another thread. 

Instead of modifying line 99, how about instead changing the "have_version" method. On line **168**, change:
`if 'None' in self.version:`
to
`if 'None' in self.version or not self.version:`

That should have the same effect in stopping your crash, and would be fine to add to pyglet.

**Comment by oscargus** [comment]
Yes, it is definitely threaded.

As long as it works (and I do not have to dig more into the rather old and badly maintained SymPy pyglet code ;-)), I'm happy. :-) I could try it to confirm, but not until next week (but I have no reason to doubt). Probably it should change order though:

`if not self.version or 'None' in self.version:`

**Comment by oscargus** [comment]
But it is too late to remember if self.version was an empty string or `None`...

**Comment by benmoran56** [comment]
It is an empty str, but just in caae I went ahead and pushed in the change as you suggested.

If you could confirm next week, that would be great. Let's leave the issue open until then.

**Comment by benmoran56** [comment]
I'm going to close this for now as it should be fixed, but please reopen it if you test again and still see issues.

Thanks!
